src/kernel/boot.c

```diff
@@ -874,20 +874,39 @@ BOOT_CODE bool_t init_freemem(word_t n_available, const p_region_t *available,
     word_t size = calculate_rootserver_size(it_v_reg, extra_bi_size_bits);
     word_t max = rootserver_max_size_bits(extra_bi_size_bits);
     for (; i >= 0; i--) {
-        word_t next = i + 1;
+        /* Invariant: both i and (i + 1) are valid indices in ndks_boot.freemem. */
+        assert(i < ARRAY_SIZE(ndks_boot.freemem) - 1);
```

Comment:
Can't we even state `assert(i < ARRAY_SIZE(ndks_boot.freemem) - 2);` here, because we enforce an empty element at the end and the "skip any empty regions" loop in line 870 will walk over that one.